sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -246,46 +267,242 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+    //
+    // For example:
+    //
+    //   SELECT col1, SUM(col2) IN (SELECT c2 FROM v1) as x
+    //   FROM v2 GROUP BY col1;
+    //
+    // The above query has this plan on entry to RewritePredicateSubquery#apply:
+    //
+    //   Aggregate [col1#28], [col1#28, sum(col2#29) IN (list#24 []) AS x#25]
+    //   :  +- LocalRelation [c2#35L]
+    //   +- LocalRelation [col1#28, col2#29]
+    //
+    // Note that the Aggregate node contains the IN-subquery and the left-hand
+    // side of the IN-subquery is an aggregate expression (sum(col2#29)).
+    //
+    // This handler transforms the above plan into the following:
+    //
+    //   Project [col1#28, sum(col2)#36L IN (list#24 []) AS x#25]
+    //   :  +- LocalRelation [c2#35L]
+    //   +- Aggregate [col1#28], [col1#28, sum(col2#29) AS sum(col2)#36L]
+    //      +- LocalRelation [col1#28, col2#29]
+    //
+    // The transformation pulled the IN-subquery up into a Project. The left-hand side of the
+    // IN-subquery is now an attribute (sum(col2)#36L) that refers to the actual aggregation
+    // which is still performed in the Aggregate node (sum(col2#29) AS sum(col2)#36L). The Unary
+    // node handler will use that attribute in the join condition (rather than the aggregate
+    // expression).
+    //
+    // If the IN-subquery is nested in a larger expression, that entire larger
+    // expression is pulled up into the Project. For example:
+    //
+    //   SELECT SUM(col2) IN (SELECT c3 FROM v1) AND SUM(col3) > -1 AS x
+    //   FROM v2;
+    //
+    // The input to RewritePredicateSubquery#apply is the following plan:
+    //
+    //   Aggregate [(sum(col2#34) IN (list#28 []) AND (sum(col3#35) > -1)) AS x#29]
+    //   :  +- LocalRelation [c3#44L]
+    //   +- LocalRelation [col2#34, col3#35]
+    //
+    // This handler transforms the plan into:
+    //
+    //   Project [(sum(col2)#45L IN (list#28 []) AND (sum(col3)#46L > -1)) AS x#29]
+    //   :  +- LocalRelation [c3#44L]
+    //   +- Aggregate [sum(col2#34) AS sum(col2)#45L, sum(col3#35) AS sum(col3)#46L]
+    //      +- LocalRelation [col2#34, col3#35]
+    //
+    // Note that the entire AND expression was pulled up into the Project, but the Aggregate
+    // node continues to perform the aggregations (but without the IN-subquery expression).
+    case a: Aggregate if exprsContainsAggregateInSubquery(a.aggregateExpressions) =>
+      // Find any interesting expressions from Aggregate.aggregateExpressions.
+      //
+      // An interesting expression is one that contains an IN-subquery whose left-hand
+      // operand contains aggregates. For example:
+      //
+      //   SELECT col1, SUM(col2) IN (SELECT c2 FROM v1)
+      //   FROM v2 GROUP BY col1;
+      //
+      // withInSubquery will be a List containing a single Alias expression:
+      //
+      //   List(sum(col2#12) IN (list#8 []) AS (...)#19)
+      val withInSubquery = a.aggregateExpressions.filter(exprContainsAggregateInSubquery(_))
```

Comment:
Once we detect such `InSubquery`, I think it's much simpler to normalize the `Aggregate` node to pull up the full result projection to a new `Project` node, instead of only rewriting the problematic `InSubquery`. This is also how `RewriteWithExpression` does it and the code is much simpler and less error-prone. We can even create a util function to reuse the code in `RewriteWithExpression`.

Reply:
BTW I think it's better to always build the query plan tree with this normalized form (`Aggregate` should only do grouping and aggregating, projection should leave to `Project`), but this is a much bigger topic.

Reply:
I am delayed in responding to review comments: I not around my laptop much until next week.

Reply:
I didn't create a util function because the `PhysicalAggregation` extractor does almost all the heavy lifting and the version of the code in `RewriteWithExpression` called `applyInternal` on the new `Aggregate` node before making it a child of the new `Project` node.